qhbmlib/inference/ebm.py

```diff
@@ -345,33 +345,45 @@ def _inner_log_partition():
   def _log_partition_forward_pass(self):
     """Returns approximation to the log partition function.
 
-    Note this simple estimator is biased.  See the paper appendix for its
-    motivation.  TODO(#216)
+    The calculation uses the uniform distribution to approximate the
+    partition function using importance sampling.  See section 11.5 of [1] for
+    details on importance sampling, in particular 11.5.2 on self-normalized
+    importance sampling.  TODO (#216): decrease variance of this estimator.
+
+    #### References
+    [1]: Murphy, Kevin P. (2023).
+         Probabilistic Machine Learning: Advanced Topics.
+         MIT Press.
     """
-    samples = self.sample(self.num_expectation_samples)
-    unique_samples, _, _ = utils.unique_bitstrings_with_counts(samples)
-    return tf.math.reduce_logsumexp(-1.0 * self.energy(unique_samples))
+    # Sample from the uniform distribution
+    samples = tfp.distributions.Bernoulli(
+        logits=tf.zeros([self.energy.num_bits])).sample(
+            self.num_expectation_samples)
+    # gamma tilde
+    unnormalized_probabilities = tf.math.exp(-1.0 * self.energy(samples))
```

Comment:
danger: this can lead to over/underflow. hence why logsumexp will do something like shifting by the max energy

https://gregorygundersen.com/blog/2020/02/09/log-sum-exp/

Reply:
More careful derivation revealed that what I'm doing here is actually just Monte Carlo integration.  Added my derivation to the docstring, which leads to the more numerically stable logsumexp you suggested